source/text/typesetter.ts

```diff
@@ -19,199 +26,273 @@ import { Label } from './label';
-        vertex.up = vec3.fromValues(0.0, glyph.extent[1], 0.0);
+        // tslint:disable-next-line:prefer-for-of
+        for (let i0 = reverse ? labelFragments.length - 1 : 0;
+            reverse ? i0 > -1 : i0 < labelFragments.length; reverse ? --i0 : ++i0) {
 
-        vertex.uvRect[0] = glyph.subTextureOrigin[0];
-        vertex.uvRect[1] = glyph.subTextureOrigin[1];
+            const fragment = labelFragments[i0];
 
-        const upperRight = vec2.create();
-        vec2.add(upperRight, glyph.subTextureOrigin, glyph.subTextureExtent);
+            if (fragment[2] === Typesetter.FragmentType.LineFeed) {
+                continue;
+            }
 
-        vertex.uvRect[2] = upperRight[0];
-        vertex.uvRect[3] = upperRight[1];
+            /* If next fragment fits as a whole, put it in. */
+
+            if (width + labelFragmentWidths[i0] < threshold) {
+                width += labelFragmentWidths[i0];
+
+                fragments.push(fragment);
+                fragmentWidths.push(labelFragmentWidths[i0]);
+                continue;
+            }
+            /* If the single delimiter didn't fit, then break. */
+            if (fragment[2] === Typesetter.FragmentType.Delimiter) {
+                break;
+            }
+
+            /* Try to cramp as many characters of the fragment (word) as possible. */
+            for (let i1 = fragment[0]; i1 < fragment[1]; ++i1) {
+                if (width + labelAdvances[i1] + labelKernings[i1] < threshold) {
+                    width += labelAdvances[i1] + labelKernings[i1];
+                    continue;
+                }
+
+                fragments.push([fragment[0], i1, fragment[2]]);
+                fragmentWidths.push(width + labelAdvances[i1]);
+                break;
+            }
+            break;
+        }
+
+        return [fragments, fragmentWidths];
     }
 
 
     /**
      * Computes origin, tangent, and up vector for every vertex of in the given range.
      * @param transform - Transformation to apply to every vertex.
-     * @param glyphs - Glyph vertices to be transformed (expected untransformed, in typesetting space).
+     * @param vertices - Glyph vertices to be transformed (expected untransformed, in typesetting space).
      * @param begin - Vertex index to start alignment at.
      * @param end - Vertex index to stop alignment at.
      */
-    protected static transformVertex(transform: mat4,
-        glyphs: GlyphVertices | undefined, begin: number, end: number): void {
-        if (glyphs === undefined || mat4.equals(transform, mat4.create())) {
+    private static transformVertex(transform: mat4,
+        vertices: GlyphVertices | undefined, begin: number, end: number): void {
+        if (vertices === undefined || mat4.equals(transform, mat4.create())) {
             return;
         }
 
         for (let i: number = begin; i < end; ++i) {
-            const v = glyphs.vertices[i];
+            const v = vertices.vertices[i];
```

Comment:
I understand the why you renamed 'glyphs' to 'vertices', but now this line looks funny :D